src/python/bot/tasks/fuzz_task.py

```diff
@@ -1254,319 +1258,391 @@ def get_strategy_distribution_from_ndb():
+    # is done on trunk build (using revision=None). Otherwise, a job definition
+    # can provide a revision to use via |APP_REVISION|.
+    if (build_manager.setup_build(environment.get_value('APP_REVISION')) and
+        environment.get_value('DATAFLOW_BUILD_BUCKET_PATH')):
+      # Some fuzzing jobs may use auxiliary builds, such as DFSan instrumented
+      # builds accompanying libFuzzer builds to enable DFT-based fuzzing.
+      build_manager.setup_trunk_build(['DATAFLOW_BUILD_BUCKET_PATH'],
+                                      'DATAFLOW')
+
+    # Check if we have an application path. If not, our build failed
+    # to setup correctly.
+    app_path = environment.get_value('APP_PATH')
+    if not app_path:
+      _track_fuzzer_run_result(self.fuzzer_name, 0, 0,
+                               FuzzErrorCode.BUILD_SETUP_FAILED)
+      return
 
-    logs.log('Upto %d' % test_number)
+    # Check if we have a bad build, i.e. one that crashes on startup.
+    # If yes, bail out.
+    logs.log('Checking for bad build.')
+    crash_revision = environment.get_value('APP_REVISION')
+    is_bad_build = testcase_manager.check_for_bad_build(self.job_type,
+                                                        crash_revision)
+    _track_build_run_result(self.job_type, crash_revision, is_bad_build)
+    if is_bad_build:
+      return
 
-    if thread_error_occurred:
-      break
+    # Data bundle directories can also have testcases which are kept in-place
+    # because of dependencies.
+    self.data_directory = setup.get_data_bundle_directory(self.fuzzer_name)
+    if not self.data_directory:
+      _track_fuzzer_run_result(self.fuzzer_name, 0, 0,
+                               FuzzErrorCode.DATA_BUNDLE_SETUP_FAILED)
+      logs.log_error(
+          'Unable to setup data bundle %s.' % fuzzer.data_bundle_name)
+      return
 
-  # Pull testcase directory to host. The testcase file contents could have been
-  # changed (by e.g. libFuzzer) and stats files could have been written.
-  if environment.is_trusted_host():
-    from bot.untrusted_runner import file_host
-    file_host.pull_testcases_from_worker()
+    engine_impl = engine.get(fuzzer.name)
+    if engine_impl:
+      crashes = self.do_engine_fuzzing(engine_impl)
+      # TODO(ochang): fill these.
+      testcase_file_paths = []
+      testcases_metadata = {}
+      fuzzer_metadata = {}
+    else:
+      fuzzer_directory = setup.get_fuzzer_directory(self.fuzzer_name)
+      fuzzer_metadata, testcase_file_paths, testcases_metadata, crashes = (
+          self.do_blackbox_fuzzing(fuzzer, fuzzer_directory, self.job_type))
 
-  # Synchronize corpus files with GCS after fuzzing
-  if sync_corpus_directory:
-    new_files = gcs_corpus.get_new_files()
-    new_files_count = len(new_files)
-    logs.log('%d new corpus files generated by fuzzer %s (job %s).' %
-             (new_files_count, fuzz_target.project_qualified_name(), job_type))
+    if crashes is None:
+      # Error occurred.
```

Comment:
log_error  or no crashes occurred.

Reply:
This can only happen when run_fuzzer failed, and we already log_error in some of those cases. 

This also matches previous behaviour, but I think we can pipe the error a bit better here, since it's not obvious what failed here looking at the code. Added a TODO.